examples/vault-iam-auth/user-data-auth-client.sh

```diff
@@ -0,0 +1,104 @@
+    sleep 10
+  done;
+
+  echo "$description failed."
+  exit $exit_status
+}
+
+# Retrying this just in case terraform is still connecting and provisioning these files
+retry "chmod +x /tmp/auth-signature-scripts/install-dependencies.sh" "Chmod'ing signature script"
+/tmp/auth-signature-scripts/install-dependencies.sh
+
+# Creating a signed request to AWS STS API with header of server ID "vault.service.consul"
+# This request is to the method GetCallerIdentity of the AWS Security Token Service, which answers the question "who am I?"
+# This script uses python's AWS SDK boto3 to get necessary credentials and sign the request
+signed_request=$(python /tmp/auth-signature-scripts/sign-request.py vault.service.consul)
+
+iam_request_url=$(echo $signed_request | jq -r .iam_request_url)
+iam_request_body=$(echo $signed_request | jq -r .iam_request_body)
+iam_request_headers=$(echo $signed_request | jq -r .iam_request_headers)
+
+# The role name necessary here is the Vault Role name, not the AWS IAM Role name
+# This variable is filled by terraform
+data=$(cat <<EOF
+{
+  "role":"${example_role_name}",
+  "iam_http_request_method": "POST",
+  "iam_request_url": "$iam_request_url",
+  "iam_request_body": "$iam_request_body",
+  "iam_request_headers": "$iam_request_headers"
+}
+EOF
+)
+
+# We send this signed request to the Vault server
+# And the Vault server will execute this request to validate this origin with AWS
+# Retry in case the vault server is still booting and unsealing
+# Or in case run-consul running on the background didn't finish yet
+retry "curl --request POST --data '$data' https://vault.service.consul:8200/v1/auth/aws/login" "Trying to login to vault"
+
+
+# If vault cli is installed we can also perform these operations with vault cli
+# The necessary VAULT_TOKEN and VAULT_ADDR environment variables have to be set
+# This assumes you have AWS credentials configured in the standard locations AWS SDKs
+# search for credentials (environment variables (), ~/.aws/credentials, IAM instance profile,
+# or ECS task role, in that order).
+# export VAULT_TOKEN=$token
+# export VAULT_ADDR=https://vault.service.consul:8200
+# vault login -method=aws header_value=vault.service.consul role=aws-role-name
+
+# Example of getting temporary credentials with iam role from instance metadata
+# The AWS session token is necessary here because these credentials are temporary
+# creds=$(curl http://169.254.169.254/latest/meta-data/iam/security-credentials/<AWS-IAM-ROLE-NAME>)
+# export AWS_ACCESS_KEY_ID=$(echo $creds | jq -r .AccessKeyId)
+# export AWS_SECRET_ACCESS_KEY=$(echo $creds | jq -r .SecretAccessKey)
+# export AWS_SESSION_TOKEN=$(echo $creds | jq -r .Token)
+
+
+# We can then use the client token from the login output once login was successful
+token=$(echo $output | jq -r .auth.client_token)
+
+# And use the token to perform operations on vault such as reading a secret
+response=$(curl \
```

Comment:
Any reason not to do the HTTP requests and most of the code above in python as well?

Reply:
* Limiting scope of what happens behind the scenes. 
* Having the two examples of login (using vault cli or http request) below each other in the same file.